dynamo/tools/connectivity.py

```diff
@@ -566,23 +603,88 @@ def neighbors(
     else:
         raise ImportError(f"nearest neighbor search method {method} is not supported")
 
-    if result_prefix != "":
-        result_prefix = result_prefix if result_prefix.endswith("_") else result_prefix + "_"
-
-    conn_key, dist_key, neigh_key = (
-        result_prefix + "connectivities",
-        result_prefix + "distances",
-        result_prefix + "neighbors",
-    )
-    adata.obsp[conn_key], adata.obsp[dist_key] = get_conn_dist_graph(knn, distances)
-
-    adata.uns[neigh_key] = {}
-    adata.uns[neigh_key]["indices"] = knn
-    adata.uns[neigh_key]["params"] = {
+    conn_key, dist_key, neighbor_key = _gen_neighbor_keys(result_prefix)
+    logger.info_insert_adata(conn_key, adata_attr="obsp")
+    logger.info_insert_adata(dist_key, adata_attr="obsp")
+    adata.obsp[dist_key], adata.obsp[conn_key] = get_conn_dist_graph(knn, distances)
+
+    logger.info_insert_adata(neighbor_key, adata_attr="uns")
+    logger.info_insert_adata(neighbor_key + ".indices", adata_attr="uns")
+    logger.info_insert_adata(neighbor_key + ".params", adata_attr="uns")
+    adata.uns[neighbor_key] = {}
+    adata.uns[neighbor_key]["indices"] = knn
+    adata.uns[neighbor_key]["params"] = {
         "n_neighbors": n_neighbors,
         "method": method,
         "metric": metric,
         "n_pcs": n_pca_components,
     }
 
     return adata
+
+
+def check_neighbors_completeness(
+    adata: AnnData, conn_key="connectivities", dist_key="distances", result_prefix=""
+) -> bool:
+    """Check if neighbor graph in adata is valid.
+
+    Parameters
+    ----------
+    adata : AnnData
+    conn_key : str, optional
+        connectivity key, by default "connectivities"
+    dist_key : str, optional
+        distance key, by default "distances"
+    result_prefix : str, optional
+        The result prefix in adata.uns for neighbor graph related data, by default ""
+
+    Returns
+    -------
+    bool
+        [description]
+    """
+    is_valid = True
+    conn_key, dist_key, neighbor_key = _gen_neighbor_keys(result_prefix)
+    keys = [conn_key, dist_key, neighbor_key]
+
+    # Old anndata version version
+    # conn_mat = adata.uns[neighbor_key]["connectivities"]
+    # dist_mat = adata.uns[neighbor_key]["distances"]
+
+    if (conn_key not in adata.obsp) or (dist_key not in adata.obsp) or ("indices" not in adata.uns[neighbor_key]):
+        return False
+    # New anndata stores connectivities and distances in obsp
+    conn_mat = adata.obsp[conn_key]
+    dist_mat = adata.obsp[dist_key]
+    n_obs = len(adata)
+
+    # check if connection matrix and distance matrix shapes are compatible with adata shape
+    is_valid = is_valid and tuple(conn_mat.shape) == tuple([n_obs, n_obs])
+    is_valid = is_valid and tuple(dist_mat.shape) == tuple([n_obs, n_obs])
+
+    # check if indices in nearest neighbor matrix are valid
```

Comment:
my suggestion here is that you should check whether any cell has no connections. basically you calculate the row and column sum of the connectivity/distance matrix and if an element has 0 row and column sum, then the graph needs to be recomputated. 

Also, add an argument (and may set it to be True by default), if either row or column sum is 0, recompute the neighbor graph. 

The row sum may correspond to the total outgoing edge weights  (total input edge weights) of the nearest neighbor graph.

Reply:
Should a valid neighbor graph only have nonzero row sum (connectivity sum=k) and any column sum (not nearest to any point)?

Reply:
by default, the row corresponds to output, so a valid graph can be simply defined as having no nonzero row sums